cluster/shard_writer.go

```diff
@@ -43,7 +43,9 @@ func (w *ShardWriter) WriteShard(shardID, ownerID uint64, points []tsdb.Point) e
 	if !ok {
 		panic("wrong connection type")
 	}
-	defer conn.Close() // return to pool
+	defer func(conn net.Conn) {
```

Comment:
I see -- I knew about the gotcha with goroutines, I didn't realise it was also a potential gotcha with defers. Thanks @corylanou 

So I guess the fundamental question is what happened to `conn` after the call to `defer` (in the original code) that meant it had to be changed as @jwilder has done?